cli/src/commands/extensions/api.rs

```diff
@@ -0,0 +1,257 @@
+
+use crate::commands::parse::{get_packages_from_lockfile, LOCKFILE_PARSERS};
+use crate::config::{get_current_project, Config};
+use crate::lockfiles::Parse;
+use crate::{api::PhylumApi, auth::UserInfo};
+
+use anyhow::{anyhow, Context, Error, Result};
+use deno_core::parking_lot::{MappedMutexGuard, Mutex, MutexGuard};
+use deno_core::{op, OpDecl, OpState};
+use futures::future::BoxFuture;
+use once_cell::sync::Lazy;
+use phylum_types::types::auth::{AccessToken, RefreshToken};
+use phylum_types::types::common::{JobId, ProjectId};
+use phylum_types::types::job::JobStatusResponse;
+use phylum_types::types::package::{
+    Package, PackageDescriptor, PackageStatusExtended, PackageType,
+};
+use phylum_types::types::project::ProjectDetailsResponse;
+
+/// Holds either an unawaited, boxed `Future`, or the result of awaiting the future.
+/// In other words, it is a lazy evaluator for the `Future`.
+pub(crate) enum OnceFuture<T: Unpin> {
+    Future(BoxFuture<'static, T>),
+    Awaited(T),
+}
+
+impl<T: Unpin> OnceFuture<T> {
+    fn new(t: BoxFuture<'static, T>) -> Self {
+        OnceFuture::Future(t)
+    }
+
+    async fn get(&mut self) -> &T {
+        match *self {
+            OnceFuture::Future(ref mut t) => {
+                *self = OnceFuture::Awaited(t.await);
+                match *self {
+                    OnceFuture::Future(..) => unreachable!(),
+                    OnceFuture::Awaited(ref mut t) => t,
+                }
+            }
+            OnceFuture::Awaited(ref mut t) => t,
+        }
+    }
+}
+
+/// Opaquely encapsulates the extension state.
+pub struct ExtensionState(OnceFuture<Result<PhylumApi>>);
+
+impl ExtensionState {
+    /// Borrow an instance of `ExtensionState` from a reference to Deno's
+    /// `OpState` state handler, and retrieve a reference to the inner state.
+    /// The references lives as long as `&mut OpState`.
+    async fn borrow_from(state: &mut OpState) -> Result<&PhylumApi> {
+        state
+            .borrow_mut::<ExtensionState>()
+            .0
+            .get()
+            .await
+            .as_ref()
+            .map_err(|e| anyhow!("{:?}", e))
+    }
+}
+
+impl From<BoxFuture<'static, Result<PhylumApi>>> for ExtensionState {
+    fn from(t: BoxFuture<'static, Result<PhylumApi>>) -> Self {
+        Self(OnceFuture::new(t))
+    }
+}
+
+//
+// Extension API functions
+// These functions need not be public, as Deno's declarations (`::decl()`) cloak
+// them in a data structure that is consumed by the runtime extension builder.
+//
+
+/// Analyze a lockfile.
+/// Equivalent to `phylum analyze`.
+#[op]
+async fn analyze(
+    state: Rc<RefCell<OpState>>,
+    lockfile: &str,
+    project: Option<&str>,
+    group: Option<&str>,
+) -> Result<ProjectId> {
+    // The general pattern for accessing state in extensions API functions is:
+    // - borrow from Deno's state handler
+    // - pin the borrowed reference guard
+    // - extract `Result<&PhylumApi>` from ExtensionState
```

Comment:
This seems like a strange place for this comment. What is the goal of documenting this here and how is it supposed to help someone writing new stuff?

I'd imagine most people would just copy/paste some random small excerpt from another extension API anyway, so I don't see why the steps need to be outlined in a comment here.

Reply:
I think there is value in documenting this behavior somewhere, as `RefCell`s and pinning aren't mechanisms everyone is familiar with. Probably moving this above the functions and expanding on it would be better.

Reply:
I did move the comment outside of that API and above all of the `#[op]` annotated methods, as a general development guide (I purposefully avoided making that a doc comment as it's solely intended for people who will write new ops). `RefCell` and `Pin` are mechanisms that are tightly bound to Rust itself and require a nontrivial amount of time to figure out (or at least, they did for me), so I feel that a few lines explaining the design choice are warranted there, so that future maintainers won't need to spend time figuring those out. Even better would be wrapping everything in a "smart pointer" of sorts that will be able to control the lifetime of the held reference.